packages/dht/src/connection/ManagedConnection.ts

```diff
@@ -0,0 +1,398 @@
+    ) {
+        super()
+        this.objectId = ManagedConnection.objectCounter
+        ManagedConnection.objectCounter++
+
+        this.send = this.send.bind(this)
+
+        this.ownPeerDescriptor = ownPeerDescriptor
+        this.protocolVersion = protocolVersion
+        this.outgoingConnection = outgoingConnection
+        this.incomingConnection = incomingConnection
+        this.connectionType = connectionType
+        this.connectionId = new ConnectionID()
+
+        this.onDisconnected = this.onDisconnected.bind(this)
+
+        logger.trace('creating ManagedConnection of type: ' + connectionType + ' objectId: ' + this.objectId)
+        if (incomingConnection && outgoingConnection) {
+            throw new Err.IllegalArguments('Managed connection constructor only accepts either an incoming connection OR a outgoing connection')
+        }
+
+        if (outgoingConnection) {
+            this.handshaker = new Handshaker(this.ownPeerDescriptor, this.protocolVersion, outgoingConnection)
+
+            this.handshaker.once('handshakeFailed', (errorMessage) => {
+                logger.trace('IL handshake failed for outgoing connection ' + errorMessage + ' ' +
+                    this.ownPeerDescriptor.nodeName + ', ' + this.peerDescriptor?.nodeName + ' objectid: ' + this.objectId
+                    + ' outputBuffer.length: ' + this.outputBuffer.length)
+                logger.trace(' ' + this.ownPeerDescriptor.nodeName + ', ' + this.peerDescriptor?.nodeName + ' emitting handshakeFailed')
+                this.emit('handshakeFailed')
+            })
+
+            this.handshaker.on('handshakeCompleted', (peerDescriptor: PeerDescriptor) => {
+                logger.trace('handshake completed for outgoing connection ' + this.ownPeerDescriptor.nodeName +
+                    ', ' + this.peerDescriptor?.nodeName + ' objectid: ' + this.objectId
+                    + ' outputBuffer.length: ' + this.outputBuffer.length)
+
+                this.attachImplementation(outgoingConnection!)
+                this.onHandshakeCompleted(peerDescriptor)
+            })
+
+            outgoingConnection.once('connected', () => {
+                this.handshaker!.sendHandshakeRequest()
+                this.emit('connected')
+            })
+            outgoingConnection.once('disconnected', this.onDisconnected)
+
+        } else {
+            if (incomingConnection) {
+                this.handshaker = new Handshaker(this.ownPeerDescriptor, this.protocolVersion, incomingConnection!)
+                this.handshaker.on('handshakeRequest', (peerDescriptor: PeerDescriptor) => {
+                    this.setPeerDescriptor(peerDescriptor)
+                    this.emit('handshakeRequest', peerDescriptor)
+                })
+
+                incomingConnection.on('disconnected', this.onDisconnected)
+
+            }
+        }
+    }
+
+    // eventemitter3 does not implement the standard 'newListener' event, so we need to override
```

Comment:
Would it make sense to use https://github.com/streamr-dev/network-monorepo/blob/streamr-1.0/packages/utils/src/ObservableEventEmitter.ts#L13